src/transformers/models/distilbert/modeling_distilbert.py

```diff
@@ -244,20 +323,62 @@ def forward(self, x, attn_mask=None, head_mask=None, output_attentions=False):
             sa_weights: torch.tensor(bs, n_heads, seq_length, seq_length) The attention weights ffn_output:
             torch.tensor(bs, seq_length, dim) The output of the transformer block contextualization.
         """
+
+        # decoder uni-directional self-attention cached key/values tuple is at positions 1,2
+        self_attn_past_key_value = past_key_values[:2] if past_key_values is not None else None
         # Self-Attention
-        sa_output = self.attention(
+        self_attention_outputs = self.attention(
             query=x,
             key=x,
             value=x,
             mask=attn_mask,
             head_mask=head_mask,
             output_attentions=output_attentions,
+            past_key_values=self_attn_past_key_value,
         )
+
         if output_attentions:
-            sa_output, sa_weights = sa_output  # (bs, seq_length, dim), (bs, n_heads, seq_length, seq_length)
+            (
+                sa_output,
+                sa_weights,
+            ) = self_attention_outputs  # (bs, seq_length, dim), (bs, n_heads, seq_length, seq_length)
         else:  # To handle these `output_attentions` or `output_hidden_states` cases returning tuples
-            assert type(sa_output) == tuple
-            sa_output = sa_output[0]
+            assert type(self_attention_outputs) == tuple
+            sa_output = self_attention_outputs[0]
+
+        # if decoder, the last output is tuple of self-attn cache
+        if self.is_decoder:
+            outputs = self_attention_outputs[1:-1]
+            present_key_value = self_attention_outputs[-1]
+        else:
+            outputs = self_attention_outputs[1:]  # add self attentions if we output attention weights
+
+        cross_attn_present_key_value = None
+        if self.is_decoder and encoder_hidden_states is not None:
+            assert hasattr(
+                self, "crossattention"
+            ), f"If `encoder_hidden_states` are passed, {self} has to be instantiated with cross-attention layers by setting `config.add_cross_attention=True`"
+
+            # cross_attn cached key/values tuple is at positions 3,4 of past_key_value tuple
+            cross_attn_past_key_value = past_key_values[-2:] if past_key_values is not None else None
+            cross_attention_outputs = self.crossattention(
+                query=x,
+                key=x,
+                value=x,
```

Comment:
As commented above, would be better to refactor the attention to accept single `hidden_states` instead of `query, key, value`.